### DISCUSSION TITLE
jsonschema for persistence

### DISCUSSION THREAD

**Discussion by magick93** [statelyai/xstate#3243] [Q&A]
Hello

I'm investigating how best to persist the state of a statemachine. Related to this, is there an existing jsonschema for machines and states?

**Comment by davidkpiano** [comment]
There is: https://github.com/statelyai/xstate/blob/main/packages/core/src/machine.schema.json

**Reply by magick93** [reply]
Thanks. 

For the below, what type of items do these arrays take?

```
"entry": {
      "type": "array"
    },
"exit": {
    "type": "array"
```

**Reply by davidkpiano** [reply]
Sorry, those are underspecified... they should take in an array of action objects with `{ type: string; }`, with additional properties allowed.

**Reply by magick93** [reply]
Ok, and `action` is a string pointer to a function?
